File: sprint02/JMS/PGS_43163.cpp
#include <string>
#include <cstring>
#include <vector>

using namespace std;

int mincnt = 51, f = 0;

int CountCompare(string A, string B) {
    int temp = 0;

    for (int i = 0; i < A.length(); i++) {
        if (A[i] == B[i])
            temp++;
    }

    return temp;
}

void RecurFind(int index, vector<string> words, string target, int cnt, int visited[]) {
    visited[index] = 1;
    cnt++;

    if (words[index].compare(target) == 0) {
        f = 1;
        if (mincnt > cnt)
            mincnt = cnt;
    }
    else {
        for (int i = 0; i < words.size(); i++) {
            if (!visited[i] && CountCompare(words[i], words[index]) == words[i].length() - 1)
                RecurFind(i, words, target, cnt, visited);
        }
    }

    visited[index] = 0;
}

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    int visited[50];

    for (int i = 0; i < words.size(); i++) {
        if (CountCompare(words[i], begin) == begin.length() - 1) {
            memset(visited, 0, sizeof(visited));
            f = 0;

            RecurFind(i, words, target, 0, visited);

            if (f)
                answer = mincnt;
        }
    }

    return answer;
}

File: sprint02/KKW/PGS_43163.py
from collections import deque
def solution(begin, target, words):
    answer = 0
    d = deque()
    d.append([begin, 0])
    visited = [0 for _ in range(len(words))]
    while d:
        start, cnt = d.popleft()
        if start == target:
            answer = cnt
            break
        for i in range(len(words)):
            tmpcnt = 0
            if not visited[i]:
                for j in range(len(start)):
                    if start[j] != words[i][j]: tmpcnt += 1
                if tmpcnt == 1:
                    d.append([words[i], cnt + 1])
                    visited[i] = 1

    return answer